core: parse `004 RPL_MYINFO` and `005 RPL_ISUPPORT`

HumorBaby opened the pull request:
Adds parsing of two additional server events: `RPL_MYINFO` and `RPL_ISUPPORT`.

- #1082/#1496 /#1305
  - `RPL_ISUPPORT` parameters are parsed into ~`bot.server._isupport`~ `bot.server_isupport` and are accessed from ~`bot.server.PARAMETER_NAME`~ `bot.server_isupport.PARAMETER_NAME` or `bot.server_isupport['PARAMETER_NAME']` (uppercase ~; e.g., `bot.server.NAMESX`, `bot.server.TARGMAX`, `bot.server.NICKLEN`~)
- #1532 
  - IRC network name is at ~`bot.server.network`~ `bot.server_network` (technically, also accessbile as ~`bot.server.NETWORK`~ `bot.server_isupport.NETWORK` b/c its a ISUPPORT parameter)
  - IRC server hostname is at ~`bot.server.hostname`~ `bot.server_hostname`

**ISUPPORT Sample:**
_Triggers_:
```
@time=2019-04-01T18:28:13.877Z :verne.freenode.net 005 hb-testbot CHANTYPES=# EXCEPTS INVEX CHANMODES=eIbq,k,flj,CFLMPQScgimnprstz CHANLIMIT=#:120 PREFIX=(ov)@+ MAXLIST=bqeI:100 MODES=4 NETWORK=freenode STATUSMSG=@+ CALLERID=g CASEMAPPING=rfc1459 :are supported by this server
@time=2019-04-01T18:28:13.877Z :verne.freenode.net 005 hb-testbot CHARSET=ascii NICKLEN=16 CHANNELLEN=50 TOPICLEN=390 DEAF=D FNC TARGMAX=NAMES:1,LIST:1,KICK:1,WHOIS:1,PRIVMSG:4,NOTICE:4,ACCEPT:,MONITOR: EXTBAN=$,ajrxz CLIENTVER=3.0 ETRACE WHOX KNOCK :are supported by this server
@time=2019-04-01T18:28:13.878Z :verne.freenode.net 005 hb-testbot SAFELIST ELIST=CTU CPRIVMSG CNOTICE :are supported by this server
```
_Parsed_:
```
>>> json.dumps(bot.server_isupport, indent=4)  # Slightly modified for prettiness

{                                                                                                                                                                                                                                                                                                                              
    "KNOCK": true,                                                                                                                                                                                                                                                                                                             
    "MODES": 4,                                                                                                                                                                                                                                                                                                                
    "CHANTYPES": "#",                                                                                                                                                                                                                                                                                                          
    "CHARSET": "ascii",                                                                                                                                                                                                                                                                                                        
    "DEAF": "D",                                                                                                                                                                                                                                                                                                               
    "PREFIX": "(ov)@+",                                                                                                                                                                                                                                                                                                        
    "FNC": true,                                                                                                                                                                                                                                                                                                               
    "EXCEPTS": "e",                                                                                                                                                                                                                                                                                                            
    "SAFELIST": true,                                                                                                                                                                                                                                                                                                          
    "NETWORK": "freenode",                                                                                                                                                                                                                                                                                                     
    "STATUSMSG": "@+",                                                                                                                                                                                                                                                                                                         
    "ELIST": "CTU",                                                                                                                                                                                                                                                                                                            
    "TOPICLEN": 390,                                                                                                                                                                                                                                                                                                           
    "CASEMAPPING": "rfc1459",                                                                                                                                                                                                                                                                                                  
    "MAXLIST": {                                                                                                                                                                                                                                                                                                               
        "bqeI": 100                                                                                                                                                                                                                                                                                                            
    },                                                                                                                                                                                                                                                                                                                         
    "NICKLEN": 16,  # Certain parameters are cast to correct type (e.g., int)                                                                                                                                                                                                                                                                                                             
    "CNOTICE": true,                                                                                                                                                                                                                                                                                                           
    "CLIENTVER": "3.0",                                                                                                                                                                                                                                                                                                        
    "TARGMAX": {                                                                                                                                                                                                                                                                                                               
        "NOTICE": 4,                                                                                                                                                                                                                                                                                                           
        "WHOIS": 1,                                                                                                                                                                                                                                                                                                            
        "LIST": 1,  # Limits are cast to int                                                                                                                                                                                                                                                                                                           
        "ACCEPT": null,  # None = no limit                                                                                                                                                                                                                                                                                                        
        "MONITOR": null,                                                                                                                                                                                                                                                                                                       
        "PRIVMSG": 4,                                                                                                                                                                                                                                                                                                          
        "KICK": 1,                                                                                                                                                                                                                                                                                                             
        "NAMES": 1                                                                                                                                                                                                                                                                                                             
    },                                                                                                                                                                                                                                                                                                                         
    "WHOX": true,                                                                                                                                                                                                                                                                                                              
    "CHANLIMIT": {                                                                                                                                                                                                                                                                                                             
        "#": 120                                                                                                                                                                                                                                                                                                               
    },                                                                                                                                                                                                                                                                                                                         
    "CPRIVMSG": true,                                                                                                                                                                                                                                                                                                          
    "CALLERID": "g",                                                                                                                                                                                                                                                                                                           
    "INVEX": "I",                                                                                                                                                                                                                                                                                                              
    "ETRACE": true,                                                                                                                                                                                                                                                                                                            
    "CHANMODES": [                                                                                                                                                                                                                                                                                                             
        "eIbq",                                                                                                                                                                                                                                                                                                                
        "k",                                                                                                                                                                                                                                                                                                                   
        "flj",                                                                                                                                                                                                                                                                                                                 
        "CFLMPQScgimnprstz"                                                                                                                                                                                                                                                                                                    
    ],                                                                                                                                                                                                                                                                                                                         
    "CHANNELLEN": 50,                                                                                                                                                                                                                                                                                                          
    "EXTBAN": [                                                                                                                                                                                                                                                                                                                
        "$",                                                                                                                                                                                                                                                                                                                   
        "ajrxz"                                                                                                                                                                                                                                                                                                                
    ]                                                                                                                                                                                                                                                                                                                          
}
```

This PR sets the foundation for #1496, and closes #1082, closes #1532, and closes #1305.

HumorBaby commented:
Also could close #1305 --> `bot.server.NICKLEN`

HumorBaby commented:
It might be worth discussing default values for `ISUPPORT` parameters. 

For example, when should a parameter default to `None`? `bot.server._isupport` is a `defaultdict(lambda: None)`, but sub-dictionaries, like `bot.server._isupport['TARGMAX']` don't have defaults. In the later case, should `None` represent no limit?... or should it be some constant, and `None` can represent missing parameter value (if not sent from server)?

deathbybandaid commented:
I think we should keep a defaultdict of all the values that we need to use, and default values, if serverdict doesn't have the value, copy it from the defaultdict

HumorBaby commented:
:white_check_mark: Tests have arrived :tada: :

HumorBaby commented:
Force push was:
1) @deathbybandaid reminded me that `005` has two possible events that would both trigger the parsing logic. Added a verification to ensure correct event is being parsed for `RPL_ISUPPORT`.
2) Updated commit messages subjects for compliance with repo standards (I hope :grimacing:)

Exirel commented:
@HumorBaby do you have time for that PR? I kind of like the idea of the support for `004` and `005`, this could be useful for many things. :-)

HumorBaby commented:
> I kind of like the idea of the support for 004 and 005, this could be useful for many things. :-)

Let me see if I can wrap up your requested changes tonight (EST) :grin: (90% likely)

HumorBaby commented:
Here are some new commits (FINALLY!! sorry for the delay :weary:)...

In the process of addressing the review comments, I found it easier to _restart_ from the current master. Reverting certain commits and fixing/squashing the changes required to fully address the review would have been a nightmare. It was nice that in the process of redoing this PR to address the commits, I identified a good amount of bugs that were present in the previous code anyway, streamlined a lot of unnecessary code, _AANNDD_ I now :heart: sphinx-formatted documenting :tada:

Anyway, I am responding to each comment (despite being outdated), but I expect a fresh review will be easier(??). Also, here is a summary of your main concerns from the review, @Exirel:
- No more private classes/methods/constants
    - everything is now at the top of a module
- Use proper `property` decorators
- Sphinx-formatted docstrings/documentation
- Improved tests

HumorBaby commented on sopel/bot.py:
```diff
@@ -48,6 +48,38 @@ def nop(bot, cap):
         self.success = success or nop
 
 
+class _ServerInfo(object):  # Eventually this can become a Python 3 data class
```
This class has been renamed to `ServerISupport` and deals solely with ISUPPORT now.

HumorBaby commented on sopel/bot.py:
```diff
@@ -48,6 +48,38 @@ def nop(bot, cap):
+class _ServerInfo(object):  # Eventually this can become a Python 3 data class
+    """A data class for server information.
+
+    capabilities: A dict mapping supported IRCv3 capabilities to their options.
```
This specific request is no longer applicable, but the newer class should have appropriate docstrings.

HumorBaby commented on sopel/bot.py:
```diff
@@ -48,6 +48,38 @@ def nop(bot, cap):
+        as the value.
+
+        For servers that do not support IRCv3, this will be an empty
+        set.
```
Since I moved the `bot.server_capabilities` object into the  `_ServerInfo` class, I just copy-pasted the existing docstring, which has the error you reference. Since the new ISUPPORT class does not touch the `server_capabilties`, this error still exists (as it has for a long time).

HumorBaby commented on sopel/bot.py:
```diff
@@ -48,6 +48,38 @@ def nop(bot, cap):
+        self.hostname = None
+        self._isupport = collections.defaultdict(lambda: None)  # Will be falsy when parameter not in RPL_ISUPPORT
+
+    network = property(lambda self: self._isupport['NETWORK'])
```
This was moved directly under `Sopel` to be accessed as a (_properly decorated_) property: `bot.server_network`

HumorBaby commented on sopel/bot.py:
```diff
@@ -78,14 +110,9 @@ def __init__(self, config, daemon=False):
-        For example, if the server specifies the capability ``sasl=EXTERNAL``,
-        it will be here as ``{"sasl": "EXTERNAL"}``. Capabilities specified
-        without any options will have ``None`` as the value.
+        self.server = _ServerInfo()
```
`self.server` is now `self.server_isupport` since this attribute deals solely with ISUPPORT.

HumorBaby commented on sopel/coretasks.py:
```diff
@@ -162,6 +162,121 @@ def retry_join(bot, trigger):
+
+    005 RPL_ISUPPORT "<client> <1-13 tokens> :are supported by this server"
+    """
+    DEFAULT_VALUES = {
```
The number of constants just for ISUPPORT has been moved under `sopel.tools.isupport` because of how many constants (and how many lines for each constant object).

HumorBaby commented on sopel/coretasks.py:
```diff
@@ -162,6 +162,121 @@ def retry_join(bot, trigger):
+        'INVEX': 'I',
+    }
+
+    CAST_TO = {
```
:+1:

HumorBaby commented on sopel/coretasks.py:
```diff
@@ -162,6 +162,121 @@ def retry_join(bot, trigger):
+    if trigger.args[-1] != 'are supported by this server':
+        return
+
+    def parse_value(key, value):
```
No more private _things_! Also, you should find proper (hopefully) sphinx-formatted docstrings now!

HumorBaby commented on sopel/coretasks.py:
```diff
@@ -162,6 +162,121 @@ def retry_join(bot, trigger):
+    for param in parameters:
+        if '=' not in param:
+            # Simple flag (e.g,. SAFELIST)
+            val = True if param not in DEFAULT_VALUES else DEFAULT_VALUES[param]
```
Done.

HumorBaby commented on sopel/bot.py:
```diff
@@ -48,6 +48,38 @@ def nop(bot, cap):
+    network: A string with the IRC network name from RPL_ISUPPORT.
+    """
+    def __init__(self):
+        self.capabilities = {}
```
the `Sopel` class now has separate objects for ISUPPORT information, `server_hostname`, and `server_network`.

HumorBaby commented on test/test_coretasks.py:
```diff
@@ -0,0 +1,116 @@
+# coding=utf-8
```
The current master already has tests for `coretasks`, so I followed the example already set in there... hopefully, the tests are now much less complicated, more intuitive, and more explicit :grimacing:

HumorBaby commented:
So, about that _epiphany_ I had earlier: I realized that there was no reason why the ISUPPORT information needed to be accessible `dict`-like _AND_ with attributes. Access both ways would just lead to inconsistent code in the future... many other `bot` attributes are already simple dictionaries, so I updated the `ServerISupport` class to be a dictionary. This also obviated the need for a special `NOT_ADVERTISED` value to indicate an unadvertised parameter; `KeyError`s (and `... is in bot.server_isupport ...`) are so much more consistent and easier (IMO)

Thus, I added a few `tmp` commits that will be squashed (upon successful review) into the appropriate PR commits to make it seem like I never had the silly ideas that they fixed in the first place.

Exirel commented:
Let me see that. 👓

Exirel commented on test/test_coretasks.py:
```diff
@@ -108,3 +111,81 @@ def test_bot_mixed_mode_types(sopel):
+    assert sopel.server_isupport['EXTBAN'] == {'prefix': '~', 'types': 'cqnr'}
+    assert sopel.server_isupport['PREFIX'] == {'modes': 'ov', 'prefixes': '@+'}
+
+    # Negate a feature (TARGMAX)
```
This should be in its own test case.

dgw commented:
We're building this out in #1758, since @HumorBaby has been unable to devote any time to the project for a while. Thanks for the starting point / inspiration!

HumorBaby commented:
I'm sad I failed to follow through on this PR, but over-joyed that it has found such a loving and caring home in #1758 :smile_cat:
